File: mouve/NodeEditorView.cpp
#include "NodeEditorView.h"
#include <QWheelEvent>

NodeEditorView::NodeEditorView(QWidget* parent)
	: QGraphicsView(parent)
	, mZoom(1.0f)
{
	setRenderHint(QPainter::Antialiasing);
	scale(mZoom, mZoom);
	setHorizontalScrollBarPolicy(Qt::ScrollBarAsNeeded);
	setVerticalScrollBarPolicy(Qt::ScrollBarAsNeeded);
	setFocusPolicy(Qt::ClickFocus);
	setDragMode(QGraphicsView::RubberBandDrag);
}

void NodeEditorView::setZoom(float zoom)
{
	if(scene() != nullptr)
	{
		QTransform transformation;
		transformation.scale(zoom, zoom);
		setTransform(transformation);
	}
}

void NodeEditorView::mousePressEvent(QMouseEvent* event)
{
	/// xXx: For now this works not entirely like desired :)
	//if(event->button() == Qt::MiddleButton)
	//{
	//	setDragMode(QGraphicsView::ScrollHandDrag);
	//}
	//else
	//{
		QGraphicsView::mousePressEvent(event);
	//}
}

void NodeEditorView::mouseReleaseEvent(QMouseEvent* event)
{
	if(event->button() == Qt::MiddleButton)
	{
		setDragMode(QGraphicsView::RubberBandDrag);
	}
	else
	{
		QGraphicsView::mouseReleaseEvent(event);
	}
}

void NodeEditorView::wheelEvent(QWheelEvent* event)
{
	if(event->orientation() == Qt::Vertical)
	{
		mZoom += event->delta() * 0.001f;
		mZoom = qMin(qMax(mZoom, 0.3f), 3.0f);
		setZoom(mZoom);
	}
}

void NodeEditorView::contextMenuEvent(QContextMenuEvent* event)
{
	emit contextMenu(event->globalPos(), mapToScene(event->pos()));
}

void NodeEditorView::keyPressEvent(QKeyEvent* event)
{
	emit keyPress(event);
}

File: UI/ViewPrototype/NodeEditorView.py
from PySide.QtCore import *
from PySide.QtGui import *

class NodeEditorView(QGraphicsView):
	def __init__(self, parent=None):
		QGraphicsView.__init__(self, parent)

		self._zoom = 1.0
		self.setRenderHint(QPainter.Antialiasing)
		self.scale(self._zoom, self._zoom)
		self.setHorizontalScrollBarPolicy(Qt.ScrollBarAsNeeded)
		self.setVerticalScrollBarPolicy(Qt.ScrollBarAsNeeded)
		self.setFocusPolicy(Qt.ClickFocus)

		# !TODO: Make scroll wheel change it
		#self.setDragMode(QGraphicsView.ScrollHandDrag)
		self.setDragMode(QGraphicsView.RubberBandDrag)

	def wheelEvent(self, event):
		if event.orientation() == Qt.Vertical:
			self._zoom += event.delta() * 0.001
			self._zoom = min(max(self._zoom, 0.3), 3.0)
			self.setZoom(self._zoom)

	def setZoom(self, zoom):
		if self.scene() is not None:
			transformation = QTransform()
			transformation.scale(self._zoom, self._zoom)
			self.setTransform(transformation)

	def contextMenuEvent(self, event):
		self.contextMenu.emit(event.globalPos(), self.mapToScene(event.pos()))

	# Signals
	contextMenu = Signal(QPoint, QPointF)
